net/tcp/tcp_send_buffered.c

```diff
@@ -1145,6 +1168,20 @@ ssize_t psock_tcp_send(FAR struct socket *psock, FAR const void *buf,
       if (psock->s_sndcb == NULL)
         {
           psock->s_sndcb = tcp_callback_alloc(conn);
+
+#ifdef CONFIG_DEBUG_ASSERTIONS
```

Comment:
move the assertion to :
https://github.com/apache/incubator-nuttx/blob/master/net/devif/devif_callback.c#L354-L359

Reply:
This assertion is not to debug devif_callback.c. Instead the assertion is to debug only tcp_send_buffered.c. It is to make sure that not more than one callback is allocated on each psock_tcp_send() invocation within given socket instance.

Reply:
vfs will guarantee that invalid sockets will not pass here, the check is useless

Reply:
It is about another potential issue. I am suspecting the following:
In case of closing TCP connection the TCP callback is marked as null and the callback flags are zeroed (it is normal), however in some places of the code the socket is not yet marked as closed by the time, therefore the user app can call write() further. In this case one more callback may be created in psock_tcp_send(). Then the socket is finally marked as closed, however the second callback is alive and may trigger.
Thus my plan is first to confirm the root cause (I am not yet sure the potential issue really exists), then to fix the issue (if this root cause is confirmed).
Even if the root cause is not confirmed for now, in any case this debug assert is reasonable to be in the code forever to prevent similar issue in the future because of further code modifications.